### PR TITLE
Feature/pdct 1139 ensure document metadata is validated in db client

### PR DESCRIPTION
**Pull request by katybaulch**
# Description

- Validate document metadata using db_client
- Driveby validate family metadata using db_client
- Updated tests 

## Proposed version

Please select the option below that is most relevant from the list below. This
will be used to generate the next tag version name during auto-tagging.

- [ ] Skip auto-tagging
- [x] Patch
- [ ] Minor version
- [ ] Major version

Visit the [Semver website](https://semver.org/#summary) to understand the
difference between `MAJOR`, `MINOR`, and `PATCH` versions.

Notes:

- If none of these options are selected, auto-tagging will fail
- Where multiple options are selected, the most senior option ticked will be
  used -- e.g. Major > Minor > Patch
- If you are selecting the version in the list above using the textbox, make
  sure your selected option is marked `[x]` with no spaces in between the
  brackets and the `x`

## Type of change

Please select the option(s) below that are most relevant:

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] GitHub workflow update
- [ ] Documentation update
- [ ] Remove legacy code
- [ ] Dependency update

## How Has This Been Tested?

Please describe the tests that you added to verify your changes.

## Reviewer Checklist

- [ ] The PR represents a single feature (small drive-by fixes are also ok)
- [ ] The PR includes tests that are sufficient for the level of risk
- [ ] The code is sufficiently commented, particularly in hard-to-understand areas
- [ ] Any required documentation updates have been made
- [ ] Any TODOs added are captured in future tickets
- [ ] No FIXMEs remain

### CODE REVIEW

File: app/repository/document.py [hunk 1]
```diff
@@ -133,6 +134,11 @@ def _doc_to_dto(doc_query_return: ReadObj) -> DocumentReadDTO:
         created=cast(datetime, fdoc.created),
         last_modified=cast(datetime, fdoc.last_modified),
         slug=str(fdoc.slugs[0].name if len(fdoc.slugs) > 0 else ""),
+        metadata=(
+            cast(dict, fdoc.valid_metadata)
+            if fdoc.valid_metadata is not None
+            else {"role": []}
```
Comment (diversemix) [inline app/repository/document.py:140]: I wonder if we should pull out a `default_document_metadata` variable for this?

Comment (diversemix) [reply app/repository/document.py:140]: Another thought - why do we need "role" on there? Why can't it be an empty dict? Or is this just temp while we change over?

File: app/service/document.py [hunk 2]
```diff
@@ -124,6 +126,18 @@ def update(
...
+    family = family_service.get(doc.family_import_id)
+    if family is None:
+        raise ValidationError(f"Could not find family for {doc.family_import_id}")
+
+    # Validate metadata.
+    metadata_service.validate_metadata(
+        db,
+        family.corpus_import_id,
+        document.metadata,
+        EntitySpecificTaxonomyKeys.DOCUMENT.value,
```
Comment (diversemix) [inline app/service/document.py:138]: Oh no - was the enum not based on str (see SortOrder)

File: app/repository/document.py (hunk 1 above)
Comment (katybaulch) [reply app/repository/document.py:140]: Can do. Worth noting that my next PR after I've tested this on staging will be to do an UPDATE on the documents table to create a JSON agg object to populate the metadata field if it's NULL by reading from the value in the role column  - so might be wasted effort

File: app/service/metadata.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    :param str corpus_id: The corpus ID to get the taxonomy for.
+    :param TaxonomyDataEntry metadata: The metadata to validate.
+    :param Optional[str] entity_key: The entity specific taxonomy key if
+        exists, otherwise None (which represents validating Family
+        metadata).
+    :raises ValidationError: if the metadata is invalid.
+    :return None
+    """
+    try:
+        results = metadata_repo.validate_metadata(db, corpus_id, metadata, entity_key)
```
Comment (diversemix) [inline app/service/metadata.py:31]: Is it appropriate to still call this `metadata_repo` ? ... not sure ... resolve if you are happy!

File: tests/unit_tests/service/family/test_update_family_service.py [hunk 4]
```diff
@@ -373,7 +411,8 @@ def test_update_success_when_family_organisation_mismatch_with_user_org(
     assert geography_repo_mock.get_id_from_value.call_count == 1
     assert family_repo_mock.get.call_count == 2
     assert corpus_repo_mock.get_corpus_org_id.call_count == 1
-    assert corpus_repo_mock.get_taxonomy_from_corpus.call_count == 1
+    assert metadata_repo_mock.get_taxonomy_from_corpus.call_count == 1
```
Comment (diversemix) [inline tests/unit_tests/service/family/test_update_family_service.py:414]: Job for another time - but not convinced that this tests needs to confirm some of these calls :thinking:

**Review by diversemix — APPROVED** [review]
I'm happy - one or two questions tho

File: app/service/document.py (hunk 2 above)
Comment (katybaulch) [reply app/service/document.py:138]: Yes it is - The function `get_entity_specific_taxonomy` takes a str though not the enum, so I've had to do this until I can update  it in the db_client as a driveby

File: tests/unit_tests/service/family/test_update_family_service.py (hunk 4 above)
Comment (katybaulch) [reply tests/unit_tests/service/family/test_update_family_service.py:414]: Think you're right - they were originally to help debugging the tests with the mocks. Lets review as part of wider test strat discussion next week?

File: app/service/metadata.py (hunk 3 above)
Comment (katybaulch) [reply app/service/metadata.py:31]: Happy to change

File: app/repository/document.py (hunk 1 above)
Comment (diversemix) [reply app/repository/document.py:140]: yeah happy for you to leave as-is if that is the case

File: app/service/document.py (hunk 2 above)
Comment (diversemix) [reply app/service/document.py:138]: thanks

File: tests/unit_tests/service/family/test_update_family_service.py (hunk 4 above)
Comment (diversemix) [reply tests/unit_tests/service/family/test_update_family_service.py:414]: sgtm

File: app/service/metadata.py (hunk 3 above)
Comment (diversemix) [reply app/service/metadata.py:31]: Let's leave it as is ... I don't think there is a strong argument either way - so let's not fix what isn't broken :smile:

**Review by diversemix — APPROVED** [review]
Nice :+1:

**Review by diversemix — APPROVED** [review]
great

**Pull request merged**